digest/cache.go

```diff
@@ -36,13 +36,13 @@ func OpenShared() *Cache {
 
 const cacheName = "digests"
 
-func (c *Cache) Get(ctx context.Context, file paranoid.File) (ForUpload, error) {
+func (c *Cache) Get(ctx context.Context, file paranoid.File, forUpload func() ForUpload) (ForUpload, error) {
```

Comment:
It may be better to have CreateForUpload in a separate interface and we can pass either the client or a global struct that simply contains the CreateForUpload func.

Something like: 

```
type DigestType interface {
  CreateForUpload() ForUpload
}

type awsDigest struct {}

func (a *awsDigest)  CreateForUpload() ForUpload {
  return &digest.ForUploadAWS{}
}

var AWSDigest = awsDigest{}
```

Then instead of passing the forUpload func everywhere, we would pass the DigestType, i.e. AWSDigest.